Add more utility functions for job and status

jian-he opened the pull request:
Add more utility functions for job and status. 
This echos what's already available in TF operator.

@terrytangyuan 
@richardsliu 
@johnugeorge
@gaocegege 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/common/14)
<!-- Reviewable:end -->

jian-he commented:
For the UT,  since it requires other code to be available to write a full UT,   can it be added later ?

richardsliu commented:
Can we add some unit tests?

jian-he commented:
sure, let me do it

gaocegege commented on job_controller/status.go:
```diff
@@ -0,0 +1,146 @@
+	// labels for pods and servers.
+	ReplicaTypeLabel  = "replica-type"
+	ReplicaIndexLabel = "replica-index"
+	labelGroupName      = "group-name"
```
We define the label as private but it is not used in the package, should we define it as a global var or just delete it?

jian-he replied:
This is used in the newly added TestJobController

terrytangyuan commented on util/k8sutil/k8sutil.go:
```diff
@@ -121,3 +122,19 @@ func FilterPodCount(pods []*v1.Pod, phase v1.PodPhase) int32 {
 }
+
+func GetTotalReplicas(replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) int32 {
+	tfjobReplicas := int32(0)
```
Remove mentions of "tfjob" in this function

jian-he replied:
done

johnugeorge commented on job_controller/job_controller.go:
```diff
@@ -72,7 +79,7 @@ type ControllerInterface interface {
 
 	// DeleteService deletes the service
-	DeleteService(job interface{}, service *v1.Service) error
+	DeleteService(job interface{}, name string, namespace string) error
```
Can we make function signature consistent with CreateService?

gaocegege replied:
Personally, I think we do not need to compose a service object to delete it.

johnugeorge replied:
Is service argument meant to have the entire Service Object?  What is the `job` argument for?

jian-he replied:
the job argument is for the purpose of logging with job contexts such as its name
e.g  the r.Recorder.Eventf takes in the job as the argument

jian-he commented:
Pushed a new change.
Added a TestJobController which implements the common interfaces and also the associated UTs around it. 
Fixed the other comments too

@gaocegege @richardsliu @terrytangyuan @johnugeorge

terrytangyuan commented:
/lgtm

richardsliu commented on job_controller/status_test.go:
```diff
@@ -0,0 +1,66 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestIsSucceeded(t *testing.T) {
```
TestIsFailed?

jian-he replied:
added

richardsliu commented on job_controller/status_test.go:
```diff
@@ -0,0 +1,66 @@
+
+func TestUpdateJobReplicaStatuses(t *testing.T) {
+	jobStatus := v1.JobStatus{}
+	initializeReplicaStatuses(&jobStatus, "worker")
```
Add asserts here for the initialized statuses

jian-he replied:
added

richardsliu commented on job_controller/status.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// filterOutCondition returns a new slice of job conditions without conditions with the provided type.
+func filterOutCondition(conditions []common.JobCondition, condType common.JobConditionType) []common.JobCondition {
```
Is this tested?

jian-he replied:
added

richardsliu commented on job_controller/status.go:
```diff
@@ -0,0 +1,144 @@
+// setCondition updates the job to include the provided condition.
+// If the condition that we are about to add already exists
+// and has the same status and reason then we are not going to update.
+func setCondition(status *common.JobStatus, condition common.JobCondition) {
```
Should have a unit test for this.

jian-he replied:
This is called by updateJobConditions and already covered by TestUpdateJobConditions

jian-he commented:
@richardsliu check the new commit

richardsliu commented:
/lgtm
/approve
